cdap-app-fabric/src/main/java/io/cdap/cdap/internal/capability/CapabilityStatusStore.java

```diff
@@ -0,0 +1,210 @@
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.internal.capability;
+
+import com.google.gson.Gson;
+import com.google.inject.Inject;
+import io.cdap.cdap.api.dataset.lib.CloseableIterator;
+import io.cdap.cdap.spi.data.StructuredRow;
+import io.cdap.cdap.spi.data.StructuredTable;
+import io.cdap.cdap.spi.data.table.field.Field;
+import io.cdap.cdap.spi.data.table.field.Fields;
+import io.cdap.cdap.spi.data.table.field.Range;
+import io.cdap.cdap.spi.data.transaction.TransactionRunner;
+import io.cdap.cdap.spi.data.transaction.TransactionRunners;
+import io.cdap.cdap.store.StoreDefinition;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+
+/**
+ * CapabilityStatusStore which takes care of reading , writing capability status and provides additional helpful methods
+ */
+public class CapabilityStatusStore implements CapabilityReader, CapabilityWriter {
+
+  private static final Gson GSON = new Gson();
+  private final TransactionRunner transactionRunner;
+
+  @Inject
+  CapabilityStatusStore(TransactionRunner transactionRunner) {
+    this.transactionRunner = transactionRunner;
+  }
+
+  /**
+   * Return the current status for a capability. If capability is not present, throws {@link IllegalArgumentException}
+   *
+   * @param capability
+   * @return {@link CapabilityStatus}
+   */
+  public CapabilityStatus getStatus(String capability) throws IOException {
+    return TransactionRunners.run(transactionRunner, context -> {
+      StructuredTable capabilityTable = context.getTable(StoreDefinition.CapabilitiesStore.CAPABILITIES);
+      Collection<Field<?>> keyField = Collections
+        .singleton(Fields.stringField(StoreDefinition.CapabilitiesStore.NAME_FIELD, capability));
+      Collection<String> returnField = Collections.singleton(StoreDefinition.CapabilitiesStore.STATUS_FIELD);
+      Optional<StructuredRow> result = capabilityTable.read(keyField, returnField);
+      return result.map(structuredRow -> CapabilityStatus
+        .valueOf(structuredRow.getString(StoreDefinition.CapabilitiesStore.STATUS_FIELD).toUpperCase())).orElse(null);
+    }, IOException.class);
+  }
+
+  @Override
+  public boolean isEnabled(String capability) throws IOException {
+    return getStatus(capability) == CapabilityStatus.ENABLED;
+  }
+
+  @Override
+  public void checkAllEnabled(Collection<String> capabilities) throws IOException, CapabilityNotAvailableException {
+    for (String capability : capabilities) {
+      if (!isEnabled(capability)) {
```

Comment:
this is not an efficient implementation. You should be able to make one query to the storage, instead of querying them one by one.

Reply:
changed to single query.